File: packages/shared/src/types/group.ts
import { z } from 'zod'
import {
  GROUP_MAX_MEMBERS,
  GROUP_MAX_NAME_LENGTH,
  GROUP_MAX_DESCRIPTION_LENGTH,
} from '../constants/groups.js'

// Who is allowed to invite new members. Today only 'admin' is wired up;
// the column exists so we can introduce 'any_member' later without a
// migration. Enum kept small to match the current enforcement.
export const GroupInviteRule = z.enum(['admin'])
export type GroupInviteRule = z.infer<typeof GroupInviteRule>

export const GroupRole = z.enum(['admin', 'member'])
export type GroupRole = z.infer<typeof GroupRole>

export const GroupSettings = z.object({
  who_can_invite: GroupInviteRule.default('admin'),
})
export type GroupSettings = z.infer<typeof GroupSettings>

export const GroupMember = z.object({
  handle: z.string(),
  display_name: z.string().nullable(),
  role: GroupRole,
  joined_at: z.string().datetime(),
})
export type GroupMember = z.infer<typeof GroupMember>

export const Group = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string().nullable(),
  avatar_url: z.string().nullable(),
  created_by: z.string(), // handle
  settings: GroupSettings,
  member_count: z.number().int().nonnegative(),
  created_at: z.string().datetime(),
  last_message_at: z.string().datetime().nullable(),
})
export type Group = z.infer<typeof Group>

export const GroupDetail = Group.extend({
  members: z.array(GroupMember),
  your_role: GroupRole,
})
export type GroupDetail = z.infer<typeof GroupDetail>

// Handles of the initial members to seed the group with. The creator is
// always added as admin and does not need to appear here.
export const CreateGroupRequest = z.object({
  name: z.string().min(1).max(GROUP_MAX_NAME_LENGTH),
  description: z.string().max(GROUP_MAX_DESCRIPTION_LENGTH).optional(),
  avatar_url: z.string().url().optional(),
  member_handles: z.array(z.string()).max(GROUP_MAX_MEMBERS - 1).default([]),
  settings: GroupSettings.partial().optional(),
})
export type CreateGroupRequest = z.infer<typeof CreateGroupRequest>

export const UpdateGroupRequest = z.object({
  name: z.string().min(1).max(GROUP_MAX_NAME_LENGTH).optional(),
  description: z.string().max(GROUP_MAX_DESCRIPTION_LENGTH).nullable().optional(),
  avatar_url: z.string().url().nullable().optional(),
  settings: GroupSettings.partial().optional(),
})
export type UpdateGroupRequest = z.infer<typeof UpdateGroupRequest>

export const AddMemberRequest = z.object({
  handle: z.string(),
})
export type AddMemberRequest = z.infer<typeof AddMemberRequest>

// Per-member add outcome. 'joined' = auto-added (they're a contact or
// their policy is open-to-contacts); 'invited' = pending invite row
// created (they had `group_invite_policy=open` but were not a contact).
// The service returns this so the dashboard / SDK can show "added 3
// members, 2 invitations pending" without a second round-trip.
export const AddMemberResult = z.object({
  handle: z.string(),
  outcome: z.enum(['joined', 'invited', 'already_member']),
  invite_id: z.string().optional(),
})
export type AddMemberResult = z.infer<typeof AddMemberResult>

export const GroupInvitation = z.object({
  id: z.string(),
  group_id: z.string(),
  group_name: z.string(),
  group_description: z.string().nullable(),
  group_avatar_url: z.string().nullable(),
  group_member_count: z.number().int().nonnegative(),
  inviter_handle: z.string(),
  created_at: z.string().datetime(),
})
export type GroupInvitation = z.infer<typeof GroupInvitation>

File: packages/sdk-python/src/agentchat/types/group.py
"""Group + membership + system-event types."""

from __future__ import annotations

from typing import List, Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field

GroupRole = Literal["admin", "member"]
GroupInviteRule = Literal["admin"]


class _BaseModel(BaseModel):
    model_config = ConfigDict(extra="allow", populate_by_name=True)


class GroupSettings(_BaseModel):
    who_can_invite: GroupInviteRule


class GroupMember(_BaseModel):
    handle: str
    display_name: Optional[str] = None
    role: GroupRole
    joined_at: str


class Group(_BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    avatar_url: Optional[str] = None
    created_by: str
    settings: GroupSettings
    member_count: int
    created_at: str
    last_message_at: Optional[str] = None


class GroupDetail(Group):
    members: List[GroupMember]
    your_role: GroupRole


class CreateGroupRequest(_BaseModel):
    name: str
    description: Optional[str] = None
    avatar_url: Optional[str] = None
    member_handles: Optional[List[str]] = None
    settings: Optional[GroupSettings] = None


class UpdateGroupRequest(_BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    avatar_url: Optional[str] = None
    settings: Optional[GroupSettings] = None


class AddMemberRequest(_BaseModel):
    handle: str


class AddMemberResult(_BaseModel):
    handle: str
    outcome: Literal["joined", "invited", "already_member"]
    invite_id: Optional[str] = None


class GroupInvitation(_BaseModel):
    id: str
    group_id: str
    group_name: str
    group_description: Optional[str] = None
    group_avatar_url: Optional[str] = None
    group_member_count: int
    inviter_handle: str
    created_at: str


# ─── System events ────────────────────────────────────────────────────────────


class _SystemEventBase(_BaseModel):
    schema_version: Literal[1] = 1


class MemberJoinedEvent(_SystemEventBase):
    event: Literal["member_joined"]
    agent_handle: str


class MemberLeftEvent(_SystemEventBase):
    event: Literal["member_left"]
    agent_handle: str


class MemberRemovedEvent(_SystemEventBase):
    event: Literal["member_removed"]
    agent_handle: str
    actor_handle: str


class AdminPromotedEvent(_SystemEventBase):
    event: Literal["admin_promoted"]
    agent_handle: str
    actor_handle: Optional[str] = None


class AdminDemotedEvent(_SystemEventBase):
    event: Literal["admin_demoted"]
    agent_handle: str
    actor_handle: str


class NameChangedEvent(_SystemEventBase):
    event: Literal["name_changed"]
    new_name: str
    actor_handle: str


class DescriptionChangedEvent(_SystemEventBase):
    event: Literal["description_changed"]
    actor_handle: str


class AvatarChangedEvent(_SystemEventBase):
    event: Literal["avatar_changed"]
    actor_handle: str


class GroupDeletedEvent(_SystemEventBase):
    event: Literal["group_deleted"]
    actor_handle: str


GroupSystemEvent = Union[
    MemberJoinedEvent,
    MemberLeftEvent,
    MemberRemovedEvent,
    AdminPromotedEvent,
    AdminDemotedEvent,
    NameChangedEvent,
    DescriptionChangedEvent,
    AvatarChangedEvent,
    GroupDeletedEvent,
]


class DeletedGroupInfo(_BaseModel):
    """Tombstone payload attached to HTTP 410 for former members of a deleted group."""

    group_id: str
    deleted_by_handle: str
    deleted_at: str


__all__ = [
    "GroupRole",
    "GroupInviteRule",
    "GroupSettings",
    "GroupMember",
    "Group",
    "GroupDetail",
    "CreateGroupRequest",
    "UpdateGroupRequest",
    "AddMemberRequest",
    "AddMemberResult",
    "GroupInvitation",
    "GroupSystemEvent",
    "DeletedGroupInfo",
]
